src/OpenTelemetry.Api/Internal/ActivityHelperExtensions.cs

```diff
@@ -259,11 +259,37 @@ private static class ActivityTagsEnumeratorFactory<TState>
                 ActivityTagObjectsEnumerator = DictionaryEnumerator<string, object, TState>.BuildAllocationFreeForEachDelegate(
                     typeof(Activity).GetField("_tags", BindingFlags.Instance | BindingFlags.NonPublic).FieldType);
 
-            private static readonly DictionaryEnumerator<string, object, TState>.AllocationFreeForEachDelegate
-                ActivityTagsCollectionEnumerator = DictionaryEnumerator<string, object, TState>.BuildAllocationFreeForEachDelegate(typeof(ActivityTagsCollection));
-
             private static readonly DictionaryEnumerator<string, object, TState>.ForEachDelegate ForEachTagValueCallbackRef = ForEachTagValueCallback;
 
+            private static readonly DictionaryEnumerator<string, object, TState>.AllocationFreeForEachDelegate ActivityEventTagsEnumerator;
+
+            private static readonly DictionaryEnumerator<string, object, TState>.AllocationFreeForEachDelegate ActivityLinkTagsEnumerator;
+
+            static ActivityTagsEnumeratorFactory()
+            {
+                var activityEventTagsField = typeof(ActivityEvent).GetField("_tags", BindingFlags.Instance | BindingFlags.NonPublic);
```

Comment:
Since we're going through with a patch release, I think it makes sense to implement the ultimate fallback in the event things change beyond DS 7. Also could put the defense in for Activity tags as well if ever that were to change.

Reply:
Here's another area in the OTLP exporter that might risk a similar 💥 in the future https://github.com/open-telemetry/opentelemetry-dotnet/blob/28c0bccca8b5b4f51cb2f03d54478aab2ea8ec0d/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs#L300

We actually ran into this once before https://github.com/open-telemetry/opentelemetry-dotnet/issues/1854...

Reply:
> Since we're going through with a patch release, I think it makes sense to implement the ultimate fallback in the event things change beyond DS 7. Also could put the defense in for Activity tags as well if ever that were to change.

+1

Reply:
> Here's another area in the OTLP exporter that might risk a similar 💥 in the future

I actually pinged the protobuf team the other day about if they would accept a contribution to make that OTLP hack/feature part of the public API. I don't want to link it here because it is kind of off topic but the response seemed like a _maybe_? I was going to throw a PR up but then @reyang pointed out to me there is an alternative thing [protobuf-net](https://github.com/protobuf-net/protobuf-net) which seems to use built-in `List<T>` for its codegen. Might make more sense to switch libs completely. Needs a bit of investigation so I tabled it for now.

Reply:
> Needs a bit of investigation so I tabled it for now.

I think this is fine, but before pushing out the patch release I think we should decide what our strategy is here. The protobuf hack is not a problem today, but could be tomorrow.

So question is: Do we want to mitigate for this in the 1.3.1 patch release?

I think the answer is yes.

If other agree, then sounds like options include:
1. Provide a fallback in the event the reflection fails.
2. Investigate protobuf-net and consider using it.

I'd be content with option 1. It requires no further investigation and sufficiently guards users of 1.3.1 from any surprise application crashes when upgrading their protobuf/gRPC dependencies in the future.

Reply:
switching to protobuf-net, if we do it, should be part of regular minor version bump, not ".1" patch for addressing bug fix.

A fallback would be the best option here.